src/promptflow/promptflow/_sdk/_utils.py

```diff
@@ -1196,3 +1199,95 @@ def parse_otel_span_status_code(value: int) -> str:
         return "Ok"
     else:
         return "Error"
+
+
+def _generate_meta_from_file(working_dir, source_path, entry, meta_dict, exception_list):
+    with _change_working_dir(working_dir), inject_sys_path(working_dir):
+        try:
+            result = generate_flow_meta_dict_by_file(
+                path=source_path,
+                entry=entry,
+            )
+            meta_dict.update(result)
+        except Exception as e:
+            exception_list.append(str(e))
+
+
+def _generate_flow_meta(
+    flow_directory: Path,
+    source_path: str,
+    entry: str,
+    timeout: int,
+    *,
+    load_in_subprocess: bool = True,
+) -> Dict[str, dict]:
+    """Generate tool meta from files.
+
+    :param flow_directory: flow directory
+    :param tools: tool list
+    :param raise_error: whether raise error when generate meta failed
+    :param timeout: timeout for generate meta
+    :param include_errors_in_output: whether include errors in output
+    :param load_in_subprocess: whether load tool meta with subprocess to prevent system path disturb. Default is True.
+        If set to False, will load tool meta in sync mode and timeout need to be handled outside current process.
+    :return: tool meta dict
+    """
+    if load_in_subprocess:
+        # use multiprocess generate to avoid system path disturb
+        manager = multiprocessing.Manager()
+        meta_dict = manager.dict()
+        exception_list = manager.list()
+        p = multiprocessing.Process(
+            target=_generate_meta_from_file, args=(flow_directory, source_path, entry, meta_dict, exception_list)
+        )
+        p.start()
+        p.join(timeout=timeout)
+        if p.is_alive():
+            logger.warning(f"Generate meta timeout after {timeout} seconds, terminate the process.")
+            p.terminate()
+            p.join()
+    else:
+        meta_dict, exception_list = {}, []
+
+        #  There is no built-in method to forcefully stop a running thread/coroutine in Python
+        #  because abruptly stopping a thread can cause issues like resource leaks,
+        #  deadlocks, or inconsistent states.
+        #  Caller needs to handle the timeout outside current process.
+        logger.warning(
+            "Generate meta in current process and timeout won't take effect. "
+            "Please handle timeout manually outside current process."
+        )
+        _generate_meta_from_file(flow_directory, source_path, entry, meta_dict, exception_list)
```

Comment:
when will we go this path?

Reply:
Offline synced with @elliotzh , for VSCode Extension scenario, the gen meta command will run in subprocess, we don't need to create another subprocess in such case.